fdbrpc/RESTClient.actor.cpp

```diff
@@ -0,0 +1,363 @@
+
+json_spirit::mObject RESTClient::Stats::getJSON() {
+	json_spirit::mObject o;
+
+	o["host_service"] = host_service;
+	o["requests_failed"] = requests_failed;
+	o["requests_successful"] = requests_successful;
+	o["bytes_sent"] = bytes_sent;
+
+	return o;
+}
+
+RESTClient::Stats RESTClient::Stats::operator-(const Stats& rhs) {
+	Stats r(host_service);
+
+	r.requests_failed = requests_failed - rhs.requests_failed;
+	r.requests_successful = requests_successful - rhs.requests_successful;
+	r.bytes_sent = bytes_sent - rhs.bytes_sent;
+
+	return r;
+}
+
+RESTClient::RESTClient() {}
+
+RESTClient::RESTClient(std::unordered_map<std::string, int>& knobSettings) {
+	knobs.set(knobSettings);
+}
+
+void RESTClient::setKnobs(const std::unordered_map<std::string, int>& knobSettings) {
+	knobs.set(knobSettings);
+}
+
+std::unordered_map<std::string, int> RESTClient::getKnobs() const {
+	return knobs.get();
+}
+
+ACTOR Future<Reference<HTTP::Response>> doRequest_impl(Reference<RESTClient> client,
+                                                       std::string verb,
+                                                       HTTP::Headers headers,
+                                                       RESTUrl* url,
+                                                       std::set<unsigned int> successCodes) {
+	state UnsentPacketQueue content;
+	state int contentLen = url->body.size();
+
+	if (url->body.size() > 0) {
+		PacketWriter pw(content.getWriteBuffer(url->body.size()), nullptr, Unversioned());
+		pw.serializeBytes(url->body);
+	}
+
+	std::string statsKey = RESTClient::getStatsKey(url->service, url->service);
+	auto sItr = client->statsMap.find(statsKey);
+	if (sItr == client->statsMap.end()) {
+		client->statsMap.emplace(statsKey, std::make_unique<RESTClient::Stats>(statsKey));
+	}
+
+	headers["Content-Length"] = format("%d", contentLen);
+	headers["Host"] = url->host;
+
+	state int maxTries = std::min(client->knobs.request_tries, client->knobs.connect_tries);
+	state int thisTry = 1;
+	state double nextRetryDelay = 2.0;
+	state Reference<IRateControl> sendReceiveRate = makeReference<Unlimited>();
+	state double reqTimeout = (client->knobs.request_timeout_secs * 1.0) / 60;
+	state RESTConnectionPoolKey connectPoolKey = RESTConnectionPool::getConnectionPoolKey(url->host, url->service);
+	state RESTClient::Stats* statsPtr = client->statsMap[statsKey].get();
+
+	loop {
+		state Optional<Error> err;
+		state Optional<NetworkAddress> remoteAddress;
+		state bool connectionEstablished = false;
+		state Reference<HTTP::Response> r;
+
+		try {
+			// Start connecting
+			Future<RESTConnectionPool::ReusableConnection> frconn = client->conectionPool->connect(
+			    connectPoolKey, client->knobs.secure_connection, client->knobs.max_connection_life);
+
+			// Finish connecting, do request
+			state RESTConnectionPool::ReusableConnection rconn =
+			    wait(timeoutError(frconn, client->knobs.connect_timeout));
+			connectionEstablished = true;
+
```

Comment:
Looking at the difference between where this was refactored from in S3BlobStore, it looks like there is a section that does the auth signature stuff post-connection but pre-http request.

While I think that making this separate, to avoid blocking this change on a bunch of blob testing, is a good goal, we want to make sure these can eventually be refactored to be the shared without requiring a bunch of additional changes to the REST/HTTP code.

IMO we should either do the refactor now, to get the pain over with, or at least have a dependent branch of this one that refactors S3BlobStore to use this code and do some of our own minimal testing to ensure there won't be major changes required?

Reply:
Agreed; I did think about this as part of coding and below are few points:
1. The eventual goal is to make S3BlobStore code use RESTClient and part away from maintaining it own ways to do things (given it makes REST calls), however, it is better to do it in stages. 
2. S3BlobStore code refactor if possible better to be driven by someone who is familiar with nuances of the code, plus, knows how to validate the code (on a running cluster if possible) to avoid any disruption etc. 
3. Given the main goal of the refactor was to enable code sharing and unblocking use case at hand (encryption work in this case), to me the path satisfies those requirements and anything more can be incrementally added as a new PR. 
My 2 cents. Please share your thoughts.

Reply:
Sounds good. 
Agree with refactor done soon enough to allow organic growth on both codepaths causing more refactor pain later. 

For now, will leave this patch as is. 
Thanks.